### PR TITLE
service/cloudformation: New Resource and Data Source: aws_cloudformation_type

### LINKED ISSUE #12583: Add Cloudformation RegisterType
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS Cloudformation RegisterType feature is missing from the provider

### New or Affected Resource(s)

* aws_cloudformation_stack

### References

https://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_RegisterType.html

### PR DESCRIPTION
**Pull request by bflad**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #12583

Please note: The handler binary is not included to prevent adding a large file to the repository. The test skip messaging includes directions how to generate it locally, if desired.

Output from acceptance testing in AWS Commercial:

```
--- PASS: TestAccAwsCloudformationType_basic (145.29s)
--- PASS: TestAccAwsCloudformationType_disappears (160.15s)
--- PASS: TestAccAwsCloudformationType_ExecutionRoleArn (145.92s)
--- PASS: TestAccAwsCloudformationType_LoggingConfig (147.60s)

--- PASS: TestAccAwsCloudformationTypeDataSource_Arn_Private (146.49s)
--- PASS: TestAccAwsCloudformationTypeDataSource_Arn_Public (12.57s)
--- PASS: TestAccAwsCloudformationTypeDataSource_TypeName_Private (144.70s)
--- PASS: TestAccAwsCloudformationTypeDataSource_TypeName_Public (12.46s)
```

Output from acceptance testing in AWS GovCloud (US):

```
--- PASS: TestAccAwsCloudformationType_basic (157.19s)
--- PASS: TestAccAwsCloudformationType_disappears (147.02s)
--- PASS: TestAccAwsCloudformationType_ExecutionRoleArn (146.99s)
--- PASS: TestAccAwsCloudformationType_LoggingConfig (159.26s)

--- PASS: TestAccAwsCloudformationTypeDataSource_Arn_Private (147.32s)
--- PASS: TestAccAwsCloudformationTypeDataSource_Arn_Public (18.42s)
--- PASS: TestAccAwsCloudformationTypeDataSource_TypeName_Private (147.18s)
--- PASS: TestAccAwsCloudformationTypeDataSource_TypeName_Public (18.35s)
```

### CODE REVIEW

File: aws/resource_aws_cloudformation_type.go [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsCloudFormationTypeCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*AWSClient).cfconn
+
```
Comment (ewbankkit) [inline aws/resource_aws_cloudformation_type.go:134]: Suggest making local variable `typeName` (as it's a required attribute) and using in error messages.

```suggestion
	typeName := d.Get("type_name").(string)
```

File: aws/resource_aws_cloudformation_type.go [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+	input := &cloudformation.RegisterTypeInput{
+		ClientRequestToken: aws.String(resource.UniqueId()),
+	}
```
Comment (ewbankkit) [inline aws/resource_aws_cloudformation_type.go:137]:
```suggestion
		ClientRequestToken: aws.String(resource.UniqueId()),
		TypeName:           aws.String(typeName),
	}
```

File: aws/resource_aws_cloudformation_type.go [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+
+	if v, ok := d.GetOk("type"); ok {
+		input.Type = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("type_name"); ok {
+		input.TypeName = aws.String(v.(string))
+	}
+
```
Comment (ewbankkit) [inline aws/resource_aws_cloudformation_type.go:158]: Delete.

```suggestion
```

File: aws/resource_aws_cloudformation_type.go [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+	output, err := conn.RegisterTypeWithContext(ctx, input)
+
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error registering CloudFormation Type: %w", err))
```
Comment (ewbankkit) [inline aws/resource_aws_cloudformation_type.go:162]:
```suggestion
		return diag.FromErr(fmt.Errorf("error registering CloudFormation Type (%s): %w", typeName, err))
```

File: aws/resource_aws_cloudformation_type.go [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+	}
+
+	if output == nil || output.RegistrationToken == nil {
+		return diag.FromErr(fmt.Errorf("error registering CloudFormation Type: empty response"))
```
Comment (ewbankkit) [inline aws/resource_aws_cloudformation_type.go:166]:
```suggestion
		return diag.FromErr(fmt.Errorf("error registering CloudFormation Type (%s): empty response", typeName))
```

File: aws/resource_aws_cloudformation_type.go [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+	}
+
+	registrationOutput, err := waiter.TypeRegistrationProgressStatusComplete(ctx, conn, aws.StringValue(output.RegistrationToken))
+
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error waiting for CloudFormation Type registration: %w", err))
```
Comment (ewbankkit) [inline aws/resource_aws_cloudformation_type.go:172]:
```suggestion
		return diag.FromErr(fmt.Errorf("error waiting for CloudFormation Type (%s) registration: %w", typeName, err))
```

File: aws/resource_aws_cloudformation_type.go (hunk 1 above)
Comment (bflad) [reply aws/resource_aws_cloudformation_type.go]: We used to do this more, but the APIs can change over time, so then you wind up needing the other style anyways. Adjusted it though.

**Comment by ewbankkit** [comment]
Do we want an `aws_cloudformation_types` (plural) data source?

**Comment by bflad** [comment]
@ewbankkit I can submit a followup pull request for that, if desired. No direct need/ask for it at the moment.

**Comment by ewbankkit** [comment]
Verified acceptance tests:

##### Commercial

```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAwsCloudformationType_'                                             
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsCloudformationType_ -timeout 180m
=== RUN   TestAccAwsCloudformationType_basic
=== PAUSE TestAccAwsCloudformationType_basic
=== RUN   TestAccAwsCloudformationType_disappears
=== PAUSE TestAccAwsCloudformationType_disappears
=== RUN   TestAccAwsCloudformationType_ExecutionRoleArn
=== PAUSE TestAccAwsCloudformationType_ExecutionRoleArn
=== RUN   TestAccAwsCloudformationType_LoggingConfig
=== PAUSE TestAccAwsCloudformationType_LoggingConfig
=== CONT  TestAccAwsCloudformationType_basic
=== CONT  TestAccAwsCloudformationType_LoggingConfig
=== CONT  TestAccAwsCloudformationType_disappears
=== CONT  TestAccAwsCloudformationType_ExecutionRoleArn
--- PASS: TestAccAwsCloudformationType_disappears (148.11s)
--- PASS: TestAccAwsCloudformationType_ExecutionRoleArn (149.30s)
--- PASS: TestAccAwsCloudformationType_basic (152.82s)
--- PASS: TestAccAwsCloudformationType_LoggingConfig (156.07s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	160.156s
```

##### GovCloud

```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAwsCloudformationType_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsCloudformationType_ -timeout 180m
=== RUN   TestAccAwsCloudformationType_basic
=== PAUSE TestAccAwsCloudformationType_basic
=== RUN   TestAccAwsCloudformationType_disappears
=== PAUSE TestAccAwsCloudformationType_disappears
=== RUN   TestAccAwsCloudformationType_ExecutionRoleArn
=== PAUSE TestAccAwsCloudformationType_ExecutionRoleArn
=== RUN   TestAccAwsCloudformationType_LoggingConfig
=== PAUSE TestAccAwsCloudformationType_LoggingConfig
=== CONT  TestAccAwsCloudformationType_basic
=== CONT  TestAccAwsCloudformationType_LoggingConfig
=== CONT  TestAccAwsCloudformationType_ExecutionRoleArn
=== CONT  TestAccAwsCloudformationType_disappears
--- PASS: TestAccAwsCloudformationType_ExecutionRoleArn (148.50s)
--- PASS: TestAccAwsCloudformationType_disappears (149.21s)
--- PASS: TestAccAwsCloudformationType_LoggingConfig (150.82s)
--- PASS: TestAccAwsCloudformationType_basic (163.84s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	167.741s
```

**Review by ewbankkit — APPROVED** [review]
LGTM 🚀.

**Comment by ewbankkit** [comment]
Forgot verification of the data source tests 😄:

##### Commercial

```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAwsCloudformationTypeDataSource_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsCloudformationTypeDataSource_ -timeout 180m
=== RUN   TestAccAwsCloudformationTypeDataSource_Arn_Private
=== PAUSE TestAccAwsCloudformationTypeDataSource_Arn_Private
=== RUN   TestAccAwsCloudformationTypeDataSource_Arn_Public
=== PAUSE TestAccAwsCloudformationTypeDataSource_Arn_Public
=== RUN   TestAccAwsCloudformationTypeDataSource_TypeName_Private
=== PAUSE TestAccAwsCloudformationTypeDataSource_TypeName_Private
=== RUN   TestAccAwsCloudformationTypeDataSource_TypeName_Public
=== PAUSE TestAccAwsCloudformationTypeDataSource_TypeName_Public
=== CONT  TestAccAwsCloudformationTypeDataSource_Arn_Private
=== CONT  TestAccAwsCloudformationTypeDataSource_Arn_Public
=== CONT  TestAccAwsCloudformationTypeDataSource_TypeName_Private
=== CONT  TestAccAwsCloudformationTypeDataSource_TypeName_Public
--- PASS: TestAccAwsCloudformationTypeDataSource_TypeName_Public (11.37s)
--- PASS: TestAccAwsCloudformationTypeDataSource_Arn_Public (11.48s)
--- PASS: TestAccAwsCloudformationTypeDataSource_TypeName_Private (158.65s)
--- PASS: TestAccAwsCloudformationTypeDataSource_Arn_Private (169.64s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	172.999s
```

##### GovCloud

```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAwsCloudformationTypeDataSource_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsCloudformationTypeDataSource_ -timeout 180m
=== RUN   TestAccAwsCloudformationTypeDataSource_Arn_Private
=== PAUSE TestAccAwsCloudformationTypeDataSource_Arn_Private
=== RUN   TestAccAwsCloudformationTypeDataSource_Arn_Public
=== PAUSE TestAccAwsCloudformationTypeDataSource_Arn_Public
=== RUN   TestAccAwsCloudformationTypeDataSource_TypeName_Private
=== PAUSE TestAccAwsCloudformationTypeDataSource_TypeName_Private
=== RUN   TestAccAwsCloudformationTypeDataSource_TypeName_Public
=== PAUSE TestAccAwsCloudformationTypeDataSource_TypeName_Public
=== CONT  TestAccAwsCloudformationTypeDataSource_Arn_Private
=== CONT  TestAccAwsCloudformationTypeDataSource_TypeName_Public
=== CONT  TestAccAwsCloudformationTypeDataSource_TypeName_Private
=== CONT  TestAccAwsCloudformationTypeDataSource_Arn_Public
--- PASS: TestAccAwsCloudformationTypeDataSource_TypeName_Public (14.37s)
--- PASS: TestAccAwsCloudformationTypeDataSource_Arn_Public (14.53s)
--- PASS: TestAccAwsCloudformationTypeDataSource_TypeName_Private (147.57s)
--- PASS: TestAccAwsCloudformationTypeDataSource_Arn_Private (147.62s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	150.862s
```

**Pull request merged**